File: src/math/vector2.ts
import { compare } from "./number";

class Vector2 {
  public readonly x: number;
  public readonly y: number;

  public constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  public multiply(value: number): Vector2 {
    return new Vector2(this.x * value, this.y * value);
  }

  public multiplyComponents(xMultiplier: number, yMultiplier: number): Vector2 {
    return new Vector2(this.x * xMultiplier, this.y * yMultiplier);
  }

  public divide(value: number): Vector2 {
    return new Vector2(this.x / value, this.y / value);
  }

  public add(other: Vector2): Vector2 {
    return new Vector2(this.x + other.x, this.y + other.y);
  }

  public subtract(other: Vector2): Vector2 {
    return new Vector2(this.x - other.x, this.y - other.y);
  }

  public equals(other: Vector2): boolean {
    return compare(this.x, other.x) === 0 && compare(this.y, other.y) === 0;
  }
}

export default Vector2;

File: logic/src/math/vector2.rs
use num_traits::Num;

use super::compare;

#[derive(Clone, Copy, Debug, Default)]
pub struct Vector2<T: Num + Copy> {
    pub x: T,
    pub y: T,
}

impl<T: Num + Copy> Vector2<T> {
    pub fn new(x: T, y: T) -> Self {
        Self { x, y }
    }

    pub fn divide(&self, value: T) -> Self {
        Self::new(self.x / value, self.y / value)
    }

    pub fn add_vector(&self, other: &Self) -> Self {
        Self::new(self.x + other.x, self.y + other.y)
    }

    pub fn multiply_components(&self, x_multiplier: T, y_multiplier: T) -> Self {
        Self::new(self.x * x_multiplier, self.y * y_multiplier)
    }
}

impl Vector2<f32> {
    pub fn equals(&self, other: &Self) -> bool {
        compare(self.x, other.x).is_eq() && compare(self.y, other.y).is_eq()
    }
}
